### ISSUE TITLE
Rust does not follow System V Amd64 ABI calling convention for small enum values, resulting in incorrect debugger output

### ISSUE THREAD

**Issue by khuey** [rust-lang/rust#85641]
Consider the following

```rust
fn return_result_ok() -> Result<u32, u32> {
    Ok(7)
}

fn return_result_err() -> Result<u32, u32> {
    Err(7)
}

fn main() {
    let r = return_result_ok();
    println!("{:?}", r);
    let r = return_result_err();
    println!("{:?}", r);
}
```

rustc generates the following assembly (debug build)

```asm
playground::return_result_ok:
	pushq	%rax
	movl	$7, 4(%rsp)
	movl	$0, (%rsp)
	movl	(%rsp), %eax
	movl	4(%rsp), %edx
	popq	%rcx
	retq

playground::return_result_err:
	pushq	%rax
	movl	$7, 4(%rsp)
	movl	$1, (%rsp)
	movl	(%rsp), %eax
	movl	4(%rsp), %edx
	popq	%rcx
	retq
```

The standard calling convention would pass both the discriminant and the value in a single register (or, alternatively, for non-POD types in C++, a reference to it would be passed. Regardless, there's no situation in the calling convention that would split a single 8 byte value across rax and rdx like this).

Although there's no hard requirement to follow the ABI calling convention for functions internal to the program, doing this confuses debuggers. If you compile this code and use gdb to step out of `return_result_ok` or `return_result_err` with the `finish` command gdb will correctly identify the `Ok`/`Err` variant (because the discriminant is in the lower 4 bytes of `%rax` where gdb expects to find it) but it will report the `u32` value as `0` (because the value is in `%edx` where the calling convention does not put it and gdb does not expect to find it).

**Comment by klensy** [comment]
Given output is `-Copt-level=0` i guess, for `-Copt-level=1` 1.54.nightly it will be
```asm
example::return_result_ok:
        xor     eax, eax
        mov     edx, 7
        ret

example::return_result_err:
        mov     eax, 1
        mov     edx, 7
        ret
```

**Comment by cuviper** [comment]
The debuginfo *should* be the source of truth for GDB, regardless of ABI -- have you checked whether that's correct?

**Comment by khuey** [comment]
The location of function return values is not represented in DWARF at all.

**Comment by khuey** [comment]
(That's not a rustc limitation, that's true of DWARF in general.)

**Comment by cuviper** [comment]
Ah, sorry, you're right. I thought there was location info, but subprograms only have `DW_AT_type`, leaving the ABI assumed.

**Comment by khuey** [comment]
Right. Rustc doesn't follow the ABI for `Result<u32, u32>` as an argument either but that's a lot harder to notice because after the function prologue the type is reassembled into a stack slot with an appropriate location in the DWARF.

**Comment by Aaron1011** [comment]
Rust deliberately does not use a standard ABI for normal functions (i.e. those without an explicit calling convention) - it instead uses the LLVM 'fastcall' convention (or something similar to it). However, I don't think this should prevent us from being compatible with debuggers - I suspect that we're just emitting incorrect debuginfo.

**Comment by khuey** [comment]
Again, there is no "debuginfo" for the location of function return values. If you don't follow the SysV calling convention the debugger can't find the return value.

**Comment by nagisa** [comment]
We have at times contributed changes to the tooling and I would expect the tooling to grow support for what Rust does in this particular instance as well. Rust does provide a mechanism to specify a variety of calling conventions that are not compatible with SysV and the tooling should behave sanely regardless of the ABI used.

**Comment by tromey** [comment]
gdb already has some limited support for `DW_AT_calling_convention`.  This could be extended, but Rust would have to emit it.

**Comment by khuey** [comment]
> gdb already has some limited support for `DW_AT_calling_convention`. This could be extended, but Rust would have to emit it.

I think the bigger problem would be defining "whatever rustc is doing" in the DWARF spec. IMO it makes more sense for the compiler to emit a DWARF location expression for the return value somewhere.

**Comment by khuey** [comment]
I raised an issue with the DWARF spec about allowing the use of a `DW_AT_location` to specify the location of function return values. 

https://dwarfstd.org/ShowIssue.php?issue=221105.1

**Comment by tromey** [comment]
This came up again in gdb: https://sourceware.org/bugzilla/show_bug.cgi?id=30090

**Comment by dwblaikie** [comment]
Does Rust guarantee/depend on stability of LLVM's fastcall ABI - or is it really an unstable/non-ABI boundary here and the calling convention could change at any time/is only valid within the scope of a single compilation? (I don't know much about Rust - but I guess the crate boundary is where the ABI applies, and this is code within a crate/not exposed across the boundary - and so everything in a crate must be compiled with an identical compiler, or is only compiled by a single invocation of the compiler anyway, so could never mismatch?)
